### PR TITLE
feat: add support for configDir

### PR DESCRIPTION
**Pull request by ThibautMarechal**
Fixes https://github.com/qmhc/vite-plugin-dts/issues/346

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -26,8 +26,8 @@ export function normalizePath(id: string): string {
   return posix.normalize(slash(id))
 }
 
-export function resolve(...paths: string[]) {
-  return normalizePath(_resolve(...paths))
+export function resolve(root: string, path: string) {
```
Comment (ThibautMarechal) [inline src/utils.ts:29]: Honestly, I'm not sure if it's the right place, or if I should add one more helper function to do that. 
But this implementation work for my use case where I use ${configDir} in include, exclude, basePath, outDir.

Let me know if you think I should do it differently

Comment (qmhc) [reply src/utils.ts:29]: I think the general utility methods should not be modified. There should be a dedicated step to handle this feature.

**Comment by qmhc** [comment]
Would you add a real case in `examples/ts` or other dirs under `examples`?

**Comment by beaussan** [comment]
Thanks a lot for this change! I spend the whole day trying to figure out why build wasn't working with configDir, with this patch this now fixes it!

File: src/utils.ts [hunk 2]
```diff
@@ -22,6 +22,10 @@ export function slash(p: string): string {
   return p.replace(windowsSlashRE, '/')
 }
 
+export function resolveConfigDir(path: string, root = process.cwd()) {
+  return path.replace('${configDir}', root)
+}
+
```
Comment (qmhc) [inline src/utils.ts:28]: I think the `root` is better to be required.

File: src/plugin.ts [hunk 3]
```diff
@@ -285,7 +286,12 @@ export function dtsPlugin(options: PluginOptions = {}): import('vite').Plugin {
       if (!outDirs) {
         outDirs = options.outDir
           ? ensureArray(options.outDir).map(d => ensureAbsolute(d, root))
-          : [ensureAbsolute(content?.raw.compilerOptions?.outDir || 'dist', root)]
+          : [
+              ensureAbsolute(
+                resolveConfigDir(content?.raw.compilerOptions?.outDir, 'root') || 'dist',
+                root
+              )
+            ]
```
Comment (qmhc) [inline src/plugin.ts:294]: Well, here maybe some unexpected changes.

It should be:

```ts
ensureAbsolute(
  content?.raw.compilerOptions?.outDir
    ? resolveConfigDir(content.raw.compilerOptions.outDir, root)
    : 'dist',
  root
)

File: src/utils.ts (hunk 2 above)
Comment (qmhc) [reply src/utils.ts:28]: And I'm not sure whether the `${configDir}` placeholder is force at the beginning.

Comment (ThibautMarechal) [reply src/utils.ts]: I made it required. 
I don't see a case where ${configDir} would not be at the start. All the example in the typescript documentation are located at the start of the string

**Review by qmhc — APPROVED** [review]

**Pull request merged**